sklearn/metrics/pairwise.py

```diff
@@ -551,7 +565,254 @@ def cosine_distances(X, Y=None):
+    that may contain mixed types of data, using an implementation
+    of Gower formula.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    Y : optional array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    categorical_features : optional array-like, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+        This is useful when categorical atributes are represented as integer
+        values. Categorical ordinal attributes are treated as numeric, and
+        must be marked as false.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is not provided, the categorical
+        columns will be auto-detected. Columns with strings, boolean,
+        and other object types will be identified as categorical, and columns
+        with any numeric type will be identified as non categorical.
+
+    apply_scaling : optional boolean, default True
+        Indicates if the data will be scaled between 0 and 1. If false, it
+        is assumed the data is already scaled.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    References
+    ---------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    -----
+    The numeric feature ranges are determined from both X and Y.
+
+    No support for sparse matrices.
+
+    The object type and its subtypes (e.g., str), are treated as categorical
+    features.
+
+    This implementation modifies the Gower's original similarity measure in the
+    folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation returns
+    1-S.
+
+    """
+    if issparse(X) or issparse(Y):
+        raise TypeError("Sparse matrices are not supported for gower distance")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_none = Y is None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if not y_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num = \
+        _detect_categorical_features(X, categorical_features)
```

Comment:
I haven't checked `_detect_categorical_features`'s implementation, but I guess we shouldn't try to be _smart_ if the user has already provided `categorical_features`. The docstring is also not very clear on this.

Reply:
This is to detect numeric features that must be treated as categorical (There is a especial treatment for numeric categorical for missing values as None and NaN behaves different in Python). Right now In the categorical features the user just need to pass if the features are categorical or not (True, False), or the indexes of categorical features. This function decomplicates the user's life, as it avoids the user be obliged to pass other thing than a bool, to identify numeric, numeric categorical, and categorical features.